File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.cpp
class Solution {
public:
    bool containsNearbyDuplicate(vector<int>& nums, int k) {
        
int n=nums.size();
        int i=0,j=0;
        unordered_map<int,int>mp;//occurence of each number in a window
        // the adjustment k = k + 1 when k is less than the size of the nums array ensures that the sliding window size is at least k.
if(k<n) k=k+1;
else k=n;
        while(j<n){
            mp[nums[j]]++;
            if(j-i+1<k) j++;
            else if(j-i+1==k){
             if(mp.size()<k) return true;
             mp[nums[i]]--;
             if(mp[nums[i]]==0) mp.erase(nums[i]);
             i++;
             j++;
          }  
        }
        return false;
    }
};

File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.java
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        //    boolean ans= false;
        // Map<Integer,Integer> mp = new HashMap<>();
        // for(int i=0;i<nums.length;i++){
        //     int num= nums[i];
        //     if(mp.containsKey(num) && i-mp.get(num)>k) ans=false;
        //     if(mp.containsKey(num) && i-mp.get(num)<=k) return true;
        //     mp.put(num, i);
        // }
        // return ans;

                   Map<Integer,Integer> mp = new HashMap<>();
        for(int i=0;i<nums.length;i++){
        

            int num= nums[i];
            if(mp.containsKey(num)){
                int dist= i-mp.get(num);
                if(dist<=k) return true;
            }
            mp.put(num, i);
        }

        return false;
    }
}